android/titanium/src/java/org/appcelerator/titanium/util/TiUrl.java

```diff
@@ -314,36 +315,176 @@ public static String absoluteUrl(String defaultScheme, String url, String baseUr
+					hostName = hostName.replace("%25", "%");    // Restore %-encoded chars in original URL.
+					stringBuilder.append(hostName);
+				}
+				if (portNumberString != null) {
+					// Append the port number as-is.
+					stringBuilder.append(':');
+					stringBuilder.append(portNumberString);
+				}
+				uriBuilder.encodedAuthority(stringBuilder.toString());
+			}
+
+			// Copy and encode the URL's path.
+			// Example: <Scheme>://<Authority>/<Path>?<AueryParams>#<Fragment>
+			String path = uri.getEncodedPath();
+			if ((path != null) && (path.length() > 0)) {
+				// Encode the given path.
+				path = Uri.encode(path, "!$&'()*+,;=:@/");
+				path = path.replace("%25", "%");                // Restore %-encoded chars in original URL.
+				uriBuilder.encodedPath(path);
+			} else if (uri.getEncodedQuery() != null) {
+				// The URL has query params, but no path or slash. Inject a '/' before the '?'.
+				// --------------------------------------------------------------------------------
+				// This works-around a Java URL parsing bug where...
+				// > "http://test.com?test=abc/xyz"
+				// ...will be transmitted like this...
+				// > "http://test.com/xyz?test=abc/xyz"
+				// --------------------------------------------------------------------------------
+				uriBuilder.encodedPath("/");
+			}
+
+			// Copy and encode the URL's anchor tag.
+			// This is the part that proceeds the '#' part of the URL.
+			String fragment = uri.getEncodedFragment();
+			if (fragment != null) {
+				fragment = Uri.encode(fragment, "!$&'()*+,;=:@/?");
+				fragment = fragment.replace("%25", "%");        // Restore %-encoded chars in original URL.
+				uriBuilder.encodedFragment(fragment);
 			}
-			builder.encodedAuthority(encodedAuthority);
-			builder.encodedPath(Uri.encode(Uri.decode(base.getPath()), "/:@$+&=;,"));
-			return builder.build();
-		} catch (Exception e) {
+
+			// Copy and encode the URL's query parameters.
+			// Note: Do not use Uri.encode() here. It encodes URL paths according to "RFC 2396".
+			//       Query parameters must be encoded via Java's URLEncoder instead which
+			//       encodes according to the "application/x-www-form-urlencoded" MIME format.
+			String queryString = uri.getEncodedQuery();
+			if ((queryString != null) && (queryString.length() > 0)) {
+				// Encode the query parameters string.
+				queryString = URLEncoder.encode(queryString, "utf8");
+
+				// The above is overly aggressive and encodes "legal" characters. Convert them back.
+				// This is important because some servers won't accept these characters in encoded form.
+				queryString = queryString.replace("%21", "!");
+				queryString = queryString.replace("%24", "$");
+				queryString = queryString.replace("%26", "&");
+				queryString = queryString.replace("%27", "'");
+				queryString = queryString.replace("%28", "(");
+				queryString = queryString.replace("%29", ")");
+				queryString = queryString.replace("%2B", "+");
```

Comment:
I'm going to punt on this. The String.replaceAll() regex always does a string allocation, even if no substring replacements are done (I just verified this today). Versus String.replace() will return a reference to itself if no substrings were found, avoiding an unnecessary allocation. Hmm...